### ISSUE TITLE
feat: Implement IGLA-GF16 — Trinity φ-Architecture for 16MB Parameter Golf

### ISSUE THREAD

**Issue by gHashTag** [gHashTag/zig-golden-float#3]
# 🎯 IGLA-GF16: Trinity Physics → Neural Architecture

## Mission Brief

Implement the **IGLA-GF16** model (Intelligent Golden-ratio Language Architecture) — a 16MB language model where every hyperparameter is derived from Trinity φ-algebra and uses the GF16 number format from this whitepaper.

This is NOT arbitrary — it proves that GF16's own `mantissa/exponent = 9/6 = 1.5 ≈ φ`, with the delta being exactly `α_φ = 0.118034`. The format IS the physics.

---

## Architecture Specification (All numbers derived from Trinity)

```
d_model  = 144   ← Fib(12)
n_heads  =   8   ← Fib(6)
d_head   =  18   ← 144/8
d_ffn    = 233   ← Fib(13) ≈ 144×φ = 232.99
n_layers =   7   ← adjusted for 16MB limit
vocab    = 50257  ← GPT-2 BPE tied embeddings
TOTAL    ≈ 15.8MB (GF16) ✅
```

---

### Module 1: Trinity Constants (`src/trinity_constants.zig`)
- [ ] Define `PHI = 1.6180339887498948482`
- [ ] Define `ALPHA_PHI = PHI^(-3) / 2 = 0.118033988749895` ← matches `α_s(mZ)` PDG2024
- [ ] Verify `PHI² + PHI⁻² = 3.0` exactly (Trinity Identity)
- [ ] Export Fibonacci sequence array for architecture dimensions

### Module 2: GF16 Format Proof (`docs/whitepaper.md` — section addition)
- [ ] Add formal proof: `man/exp ratio = 9/6 = 1.5`, `φ - 1.5 = 0.118034 = α_φ`
- [ ] This means GF16 deviates from ideal φ-split by exactly the strong coupling constant
- [ ] Document the three-way closure: `{GF16 format, α_s coupling, LR_init} = α_φ`

### Module 3: φ-Sparse Attention with CA-mask (`src/phi_attention.zig`)
- [ ] Build Fibonacci distance mask: visible positions = `{1,2,3,5,8,13,21,34,55,89,144}`
- [ ] Sparsity: `2.15%` (11/512 per token), reduction `46.6×`
- [ ] Scale factor: `d_head^(-φ⁻¹)` instead of `sqrt(d_head)`
- [ ] CA Rule 110 pattern for mask generation

### Module 4: Trinity Weight Init (`src/trinity_init.zig`)
- [ ] 4 physics sectors:
  - `gauge`     (attn QKV):  `std = α_φ = 0.11803399`
  - `higgs`     (attn proj): `std = α_φ × φ⁻¹ = 0.07294902`
  - `lepton`    (ffn gate):  `std = α_φ × φ⁻² = 0.04508497`
  - `cosmology` (embed):     `std = α_φ × φ⁻³ = 0.02786405`

### Module 5: φ-LR Schedule (`src/phi_schedule.zig`)
- [ ] `LR(t) = α_φ · φ^(-t/τ)` where `τ = T/(φ·27) = 228.9 steps`
- [ ] Warmup: linear to `α_φ` over `Fib(7) = 21` steps
- [ ] The constant `27 = 3³ = (φ²+φ⁻²)³` from Trinity Identity

### Module 6: JEPA-T Predictor (`src/jepa_t.zig`)
- [ ] Encoder 6 layers (~8MB) + Predictor 3 layers (~0.9MB) = φ-split
- [ ] Loss in latent space: `MSE(z_pred, sg(z_tgt))` — no softmax over vocab
- [ ] Memory saving: ~30% vs standard cross-entropy

### Module 7: Benchmarks & Proofs (`benchmarks/igla_gf16_bench.zig`)
- [ ] Reproduce BENCH-004b: GF16 = 97.67% of f32 accuracy (Δ=0.00%)
- [ ] Compare: bf16 = 9.80% (−87.87% ❌)
- [ ] Export all metrics as JSON for whitepaper figures

---

## Key Proofs for Whitepaper

| # | Proof | Value | Status |
|---|-------|-------|--------|
| ① | GF16 `man/exp = 1.5`, `φ - 1.5 = α_φ` | `0.118034` | 🔲 needs code |
| ② | Trinity init std = `α_s(mZ)` PDG2024 | `Δ = 0.03σ` | 🔲 needs benchmark |
| ③ | LR_init = `α_φ` (same constant) | `0.118034` | 🔲 needs ablation |
| ④ | BENCH-004b GF16 ≈ f32 | `97.67%` | ✅ exists |
| ⑤ | Fib `d_model/d_ffn`: `144×φ = 232.99 ≈ 233` | `Δ<0.1%` | 🔲 needs verify |

---

## Acceptance Criteria

- [ ] All 7 modules implemented in Zig
- [ ] `zig build test` passes with Trinity Identity verified to `< 1e-12`
- [ ] Benchmark JSON output for all 5 proofs
- [ ] Whitepaper section added: "IGLA-GF16: Closure of φ-Algebra in Neural Architecture"
- [ ] Total model size verified ≤ 16MB in GF16 format

---

## References

- Trinity paper: 42 φ-formulas for Standard Model constants
- GF16 whitepaper: `docs/whitepaper.md` (this repo)
- BENCH-004b results: existing benchmarks
- φ² + φ⁻² = 3 (Trinity Identity — the foundation)

---

**Priority**: 🔴 CRITICAL  
**Complexity**: L (3-5 days)  
**Agent**: implement all modules, run benchmarks, update whitepaper section

**Comment by gHashTag** [comment]
## Module implementation status

| Module | File | PR | Status |
|--------|------|-----|--------|
| 1. Trinity Constants | `src/trinity_constants.zig` | #58 | Pending merge |
| 2. GF16 Format Proof | (whitepaper section) | — | Documentation only |
| 3. φ-Sparse Attention | `src/phi_attention.zig` | #60 | Pending review |
| 4. Trinity Weight Init | `src/trinity_init.zig` | #60 | Pending review |
| 5. φ-LR Schedule | `src/phi_schedule.zig` | #58, #60 | Pending merge |
| 6. JEPA-T Predictor | `src/jepa_t.zig` | #60 | Pending review |
| 7. Benchmarks & Proofs | `benches/igla_gf16_bench.zig` | #60 | Pending review |

5 of 7 code modules implemented. Module 2 is a whitepaper docs task.
